libraries/AP_SafeRTL/AP_SafeRTL.cpp

```diff
@@ -0,0 +1,727 @@
+*    before they complete which is helpful when memory is filling up and we just
+*    need to quickly identify a handful of points which can be deleted.
+*
+*    Once the algorithms have completed the simplify.complete and
+*    prune.complete flags are set to true.  The "thorough cleanup" procedure,
+*    which is run as the vehicle initiates the SafeRTL flight mode, waits for
+*    these flags to become true.  This can force the vehicle to pause for a few
+*    seconds before initiating the return journey.
+*/
+
+AP_SafeRTL::AP_SafeRTL(const AP_AHRS& ahrs, bool example_mode) :
+    _ahrs(ahrs),
+    _example_mode(example_mode)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    _simplify.bitmask.setall();
+}
+
+// initialise safe rtl including setting up background processes
+void AP_SafeRTL::init()
+{
+    // protect against repeated call to init
+    if (_path != nullptr) {
+        return;
+    }
+
+    // constrain the path length, in case the user decided to make the path unreasonably long.
+    _points_max = constrain_int16(_points_max, 0, SAFERTL_POINTS_MAX);
+
+    // check if user has disabled SafeRTL
+    if (_points_max == 0 || !is_positive(_accuracy)) {
+        return;
+    }
+
+    // allocate arrays
+    _path = (Vector3f*)calloc(_points_max, sizeof(Vector3f));
+
+    _prune.loops_max = _points_max * SAFERTL_PRUNING_LOOP_BUFFER_LEN_MULT;
+    _prune.loops = (prune_loop_t*)calloc(_prune.loops_max, sizeof(prune_loop_t));
+
+    _simplify.stack_max = _points_max * SAFERTL_SIMPLIFY_STACK_LEN_MULT;
+    _simplify.stack = (simplify_start_finish_t*)calloc(_simplify.stack_max, sizeof(simplify_start_finish_t));
+
+    // check if memory allocation failed
+    if (_path == nullptr || _prune.loops == nullptr || _simplify.stack == nullptr) {
+        log_action(SRTL_DEACTIVATED_INIT_FAILED);
+        gcs().send_text(MAV_SEVERITY_WARNING, "SafeRTL deactivated: init failed");
+        free(_path);
+        free(_prune.loops);
+        free(_simplify.stack);
+        return;
+    }
+
+    _path_points_max = _points_max;
+
+    // create semaphore
+    _path_sem = hal.util->new_semaphore();
+
+    // when running the example sketch, we want the cleanup tasks to run when we tell them to, not in the background (so that they can be timed.)
+    if (!_example_mode){
+        // register background cleanup to run in IO thread
+        hal.scheduler->register_io_process(FUNCTOR_BIND_MEMBER(&AP_SafeRTL::run_background_cleanup, void));
+    }
+}
+
+// returns number of points on the path
+uint16_t AP_SafeRTL::get_num_points() const
+{
+    return _path_points_count;
+}
+
+// get next point on the path to home, returns true on success
+bool AP_SafeRTL::pop_point(Vector3f& point)
+{
+    // check we are active
+    if (!_active) {
+        return false;
+    }
+
+    // get semaphore
+    if (!_path_sem->take_nonblocking()) {
+        log_action(SRTL_POP_FAILED_NO_SEMAPHORE);
```

Comment:
How often does this show up in logs? (SITL doesn't count given the way that SITL manages threads/background tasks at the moment)